src/Graph/GraphicalLasso.cpp

```diff
@@ -0,0 +1,194 @@
+    return svd.matrixV() * (svd.singularValues().array().abs() > tolerance)
+      .select(svd.singularValues().array().inverse(), 0)
+      .matrix().asDiagonal() * svd.matrixU().adjoint();
+}
+
+void GraphicalLasso::run(LinearRegression *model) {
+    MatrixXf X = model->getX();
+    int num_samples = X.rows();
+    int num_features = X.cols();
+    stdNormalize(X);
+    MatrixXf S = X.transpose() * X / num_samples;
+    MatrixXf W = S;
+    MatrixXf W_inv = pseudoInverse(W);
+    MatrixXf W_inv_old = W_inv;
+    float regularizer = model->getL1_reg();
+
+    for (int epoch = 0; epoch < maxIteration; epoch++) {
+        progress = static_cast<float>(epoch) / maxIteration;
+        for (int idx = 0; idx < num_features; idx++) {
+            // partition matrixes into blocks
+            vector<MatrixXf> S_blocks = partitionBlocks(S);
+            MatrixXf W_11 = partitionBlocks(W)[0];
+            MatrixXf W_inv_11 = partitionBlocks(W_inv)[0];
+            
+            // find W_11 ^ (1/2)
+            MatrixXf W_11_half = matrixRaiseToHalf(W_11); 
+            MatrixXf b = pseudoInverse(W_11_half) * S_blocks[1];
+            
+            // run linear regression with lasso
+            MatrixXf beta = fit(model, W_11_half, b);
+            MatrixXf W_12 = W_11 * beta;
+            
+            // recover W_inv (W_inv corresponds to theta in the paper)
+            float S_22 = S_blocks.back()(0, 0);
+            float W_22 = S_22 + regularizer;
+            float W_inv_22 = 1 / (W_22 - (W_12.transpose() * beta)(0, 0)); 
+            MatrixXf W_inv_12 = -W_inv_22 * beta;
+
+            W_inv = composeBlocks(W_inv_11, W_inv_12, W_inv_22);
+            W = composeBlocks(W_11, W_12, W_22);
+            S = composeBlocks(S_blocks[0], S_blocks[1], S_blocks[2](0, 0));
+        }
+        float diff = (W_inv - W_inv_old).norm();
+        if (diff < tolerance) {
+            break;
+        } else {
+            W_inv_old = W_inv;
+        }
+    }
+    for (long col = 0; col < W_inv.cols(); col++) {
+        for (long row = 0; row < W_inv.rows(); row++) {
+            W_inv(row, col) = abs(W_inv(row, col)) < 1e-5 ? 0 : 1;
+        }
+    }
+    model->updateBeta(W_inv);
+}
+
+MatrixXf GraphicalLasso::fit(LinearRegression* model, MatrixXf& X, MatrixXf& Y) {
+    model->setL1_reg(model->getL1_reg()*10);
+    model->setX(X);
+    model->setY(Y);
+    model->initBeta();
+    float residue = model->cost();
+    VectorXf grad;
+    VectorXf in;
+    long epoch = 0;
+    while (!shouldStop && epoch < maxIteration && residue > tolerance && !shouldStop) {
+        grad = model->proximal_derivative();
+        in = model->getBeta() - learningRate * grad;
+        model->updateBeta(model->proximal_operator(in, learningRate));
+        residue = model->cost();
+        epoch++;
+    }
+    return model->getBeta();
+}
+
+void GraphicalLasso::stdNormalize(MatrixXf& matrix) {
+    RowVectorXf mean = matrix.colwise().mean();
+    // unbiased estimation
```

Comment:
Let's not worry about unbiased estimation, it doesn't make big differences in practice.